sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java

```diff
@@ -311,9 +315,14 @@ private void createRunnerAndConsumersForPTransformRecursively(
       // Get finish bundle Execution Time Metrics.
       response.addAllMonitoringInfos(
           bundleProcessor.getFinishFunctionRegistry().getExecutionTimeMonitoringInfos());
-      // Extract all other MonitoringInfos other than the execution time monitoring infos.
+      // Extract MonitoringInfos that come from the metrics container registry.
       response.addAllMonitoringInfos(
           bundleProcessor.getMetricsContainerRegistry().getMonitoringInfos());
+      // Add any additional monitoring infos that the "runners" report explicitly.
+      for (ProgressRequestCallback progressRequestCallback :
+          bundleProcessor.getProgressRequestCallbacks()) {
+        response.addAllMonitoringInfos(progressRequestCallback.getMonitoringInfos());
```

Comment:
When a bundle finished, will the `progress.workRemaining` always be 0?

Reply:
Progress metrics won't appear in the final monitoring infos since there are no active elements. Any residuals will contain their size if a sized SDF element and processing URN was used.